exporters/prometheus/exporter.go

```diff
@@ -0,0 +1,185 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package prometheus // import "github.com/open-telemetry/opentelemetry-go/exporters/prometheus"
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// Exporter is a Prometheus exporter.
+type Exporter struct {
+	reader   metric.Reader
+	registry *prometheus.Registry
+}
+
+// NewPrometheusExporter returns a Prometheus exporter.
+func NewPrometheusExporter(reader metric.Reader) (*Exporter, error) {
+	e := &Exporter{
+		reader:   reader,
+		registry: prometheus.NewRegistry(),
+	}
+
+	if err := e.registry.Register(e); err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+// Describe implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	metrics, err := e.reader.Collect(context.TODO())
+	if err != nil {
+		otel.Handle(err)
+	}
+	for _, metricData := range getMetricData(metrics) {
+		ch <- metricData.description
+	}
+}
+
+// Collect implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	metrics, err := e.reader.Collect(context.TODO())
+	if err != nil {
+		otel.Handle(err)
+	}
+
+	for _, metricData := range getMetricData(metrics) {
+		if metricData.valueType == prometheus.UntypedValue {
+			m, err := prometheus.NewConstHistogram(metricData.description, metricData.histogramCount, metricData.histogramSum, metricData.histogramBuckets, metricData.attributeValues...)
+			if err != nil {
+				otel.Handle(err)
+			}
+			ch <- m
+		} else {
+			m, err := prometheus.NewConstMetric(metricData.description, metricData.valueType, metricData.value, metricData.attributeValues...)
+			if err != nil {
+				otel.Handle(err)
+			}
+			ch <- m
+		}
+	}
+}
+
+// metricData holds the metadata as well as values for individual data points.
+type metricData struct {
+	name             string
```

Comment:
Note that the metric name should include the unit as a suffix, but should come before a `_total` suffix on counters.

[Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#metric-metadata-1)